nautobot/extras/context_managers.py

```diff
@@ -152,3 +159,57 @@ def web_request_context(
         for object_change in ObjectChange.objects.filter(request_id=change_context.change_id).iterator():
             enqueue_job_hooks(object_change)
             enqueue_webhooks(object_change)
+
+
+@contextmanager
+def object_changelogs_bulk_operation(objs, user, action):
+    def get_objectchange_for_instance(instance, context_state, action):
+        objectchange = instance.to_objectchange(action)
+        objectchange.user = user
+        objectchange.request_id = context_state.change_id
+        objectchange.change_context = context_state.context
+        objectchange.change_context_detail = context_state.context_detail[:CHANGELOG_MAX_CHANGE_CONTEXT_DETAIL]
+        return objectchange
+
+    with transaction.atomic():
+        # Disable automatic ChangeLog creation on signal to enhance performance.
+        # Multiple object updates lead to excessive database calls.
+        # Instead, we'll manually manage a bulk ChangeLog creation post-update.
+        prev_state = change_context_state.set(None)
+
+        if action == ObjectChangeActionChoices.ACTION_DELETE:
+            associations_to_delete = set()
+            notes_to_delete = set()
+            objectchange_entries = []
+            for instance in objs:
+                if isinstance(instance, BaseModel):
+                    associations = ContactAssociation.objects.filter(
+                        associated_object_type=ContentType.objects.get_for_model(type(instance)),
+                        associated_object_id=instance.pk,
+                    ).values_list("pk", flat=True)
+                    associations_to_delete.update(associations)
+
+                if hasattr(instance, "notes") and isinstance(instance.notes, NotesQuerySet):
+                    notes = instance.notes.values_list("pk", flat=True)
+                    notes_to_delete.update(notes)
+                if hasattr(instance, "to_objectchange"):
+                    objectchange = get_objectchange_for_instance(instance, prev_state.old_value, action)
+                    objectchange_entries.append(objectchange)
+            if associations_to_delete:
+                ContactAssociation.objects.filter(pk__in=associations_to_delete).delete()
+            if notes_to_delete:
+                Note.objects.filter(pk__in=notes_to_delete).delete()
+            if objectchange_entries:
+                ObjectChange.objects.bulk_create(objectchange_entries, batch_size=1000)
+        try:
+            yield
+            if action == ObjectChangeActionChoices.ACTION_UPDATE and hasattr(objs[0], "to_objectchange"):
+                objectchange_entries = []
+                for instance in objs:
+                    objectchange = get_objectchange_for_instance(instance, prev_state.old_value, action)
+                    objectchange_entries.append(objectchange)
+                ObjectChange.objects.bulk_create(objectchange_entries, batch_size=1000)
```

Comment:
I'm wondering if there's an alternate way we could implement this that would allow for a mixture of actions (CRUD) within a single atomic transaction using this wrapper (for example, for SSOT). Conceptually:

- Initialize `objectchange_entries = []`
- Disconnect the usual `_handle_changed_object` and `_handle_deleted_object` signals
- Connect alternate signals that do basically the same thing, but stick the constructed ObjectChange objects in `objectchange_entries` instead of immediately saving them to the DB
- Yield
- ObjectChange.objects.bulk_create(objectchange_entries)
- Reconnect original signals

Thinking even more outside the box, maybe `atomic` could become an option to the `change_logging()` context manager that would automatically invoke this alternate approach when set?

Reply:
As discussed earlier, we may not be able to use this approach because we could become out of sync in this kind of scenario:
```py
with change_logging(change_context):
    with transaction.atomic():
        Namespace.objects.create(name="foo")
        raise Exception("rollback that change!")

    with transaction.atomic():
        Namespace.objects.create(name="bar")
 ```

The above code would create an object change for Namespace `foo` even though it was rolled back

Reply:
I wonder if we could just do the `flush` on `transaction.on_commit`?

Reply:
It doesn't look like we're going to be able to go with the generic solution because we don't have a way to clear the queue on transaction rollback.